refact: Using bouncycastle/bcprov instead caesium

rafaeldelboni opened the pull request:
As suggested in this thread:
https://github.com/discord/discord-api-docs/discussions/3413

JohnnyJayJay commented on src/ring_discord_auth/core.clj:
```diff
@@ -24,49 +25,10 @@
-                   out-ev (eval out)]]
-         `(is (~equals (~f ~@in) ~out)))))
-
-(defn hex->bytes
-  "Converts the given string representing a hexadecimal number to a byte array.
-
-  Each byte in the resulting array comes from 2 digits in the string.
-  If the string cannot be converted, returns `nil`"
-  {:test (examples hex->bytes Arrays/equals
-           ["68c252fa74eb9b97" #_=> (byte-array [0x68 0xc2 0x52 0xfa 0x74 0xeb 0x9b 0x97])]
-           ["123" #_=> (byte-array [0x12 0x03])]
-           ["garbage" #_=> nil])}
-  ^bytes [^String hex-str]
-  (let [len (count hex-str)
-
-        result (byte-array (quot (inc len) 2))]
-    (try
-      (doseq [[i hex-part] (map-indexed vector (map (partial apply str) (partition-all 2 hex-str)))]
-        (aset result i (unchecked-byte (Short/parseShort hex-part 16))))
-      result
-      (catch NumberFormatException _ nil))))
-
```
I'm not sure why you removed all previous byte-handling logic. Those parts do not need to be and should not be changed for the replacement of the verification library.

rafaeldelboni replied:
Yeah I just replaced it with all the functions I was using on my side here, but I can understand your functions and reuse them, no probs.

JohnnyJayJay commented on src/ring_discord_auth/core.clj:
```diff
@@ -24,49 +25,10 @@
-
 (defn read-all-bytes
   "Reads all bytes from either an `InputStream` or a `ByteBuffer`.
-
```
An empty line after the docstring summary is conventional and should be kept

rafaeldelboni replied:
:+1:

JohnnyJayJay commented on src/ring_discord_auth/core.clj:
```diff
@@ -24,49 +25,10 @@
-
   If an `InputStream` is provided, it will be consumed, but not closed.
   Returns its result as a *new* byte array."
-  {:test (examples read-all-bytes Arrays/equals
-           [(ByteBuffer/allocate 0) #_=> (byte-array 0)]
-           [(io/input-stream (byte-array [0x12 0xab 0x4f])) #_=> (byte-array [0x12 0xab 0x4f])]
-           [(ByteBuffer/wrap (byte-array [0x67 0x2a 0x4b 0x23 0x5c]) 1 2) #_=> (byte-array [0x2a 0x4b])])}
```
I see no point in removing the examples - even if you don't use them as tests, they still contribute to documentation

rafaeldelboni replied:
I just thought they are very noisy to have in the middle of actual code, instead of then I prefer writing unit tests as examples, but you said something I didn't thought about the libs documentation, I will but then back.
:+1:

JohnnyJayJay commented on src/ring_discord_auth/core.clj:
```diff
@@ -81,54 +43,6 @@
                  (.get ^ByteBuffer input result)
                  result)))
 
-(defn encode
-  "Encodes the given string to a byte array using the given charset/encoding.
-
-  Returns `nil` if the charset is not available or if it doesn't support encoding."
-  {:test (examples encode Arrays/equals
-           ["Hello, world!", "utf8" #_=> (byte-array [0x48 0x65 0x6c 0x6c 0x6f 0x2c 0x20 0x77 0x6f 0x72 0x6c 0x64 0x21])]
-           ["Another one", "unknown?" #_=> nil])}
-  ^bytes [^String str ^String charset-name]
-  (if-let-all [^Charset cs (try (Charset/forName charset-name) (catch UnsupportedCharsetException _ nil) (catch IllegalCharsetNameException _ nil))
-               ^CharsetEncoder encoder (try (.newEncoder cs) (catch UnsupportedOperationException _ nil))]
-    (when (.canEncode encoder str)
-      (read-all-bytes (.encode encoder (CharBuffer/wrap str))))))
```
This function should not be removed either. `(.getBytes string "utf8")` is less versatile (the character encoding should not be hard coded) and not sound.

rafaeldelboni replied:
As I said, I just replaced it with all the functions I was using on my side here, but I can understand your functions and reuse them.

JohnnyJayJay commented on src/ring_discord_auth/core.clj:
```diff
@@ -145,24 +59,24 @@
+                    timestamp-hex timestamp
+                    signature-hex signature
+                    body (slurp raw-body)]
+                   (if (validation/verify-request public-key-hex timestamp-hex body signature-hex)
+                     (handler (assoc request :body (ByteArrayInputStream. raw-body)))
+                     {:status 401 :headers default-headers :body "Signature was not authentic."})
+                   {:status 400 :headers default-headers :body "Missing body, signature or timestamp."})
```
formatting issues, see the way it was before

rafaeldelboni replied:
I need to add clj-kondo rules here for your the custom macro :+1:

JohnnyJayJay commented on src/ring_discord_auth/core.clj:
```diff
@@ -81,54 +43,6 @@
-    (when (.canEncode encoder str)
-      (read-all-bytes (.encode encoder (CharBuffer/wrap str))))))
-
-(defn authentic?
-  "Checks whether a signature is authentic, given a message and a public key.
-
-  This function has 2 arities: a general purpose one that simply delegates to a crypto library and one tailored to what Discord provides.
-
-  The general purpose arity takes a signature, a message and a public key (all of which must be byte arrays) and checks the authenticity.
-
-  The other arity takes a signature (byte array or hex string), public key (byte array or hex string), body (byte array or string), timestamp (byte array or string).
-  It combines timestamp and body and uses that as the message.
-
-  Returns `true` if the message is authentic, `false` if not."
-  ([signature-bytes message-bytes public-key-bytes]
-   (try
-     (sign/verify signature-bytes message-bytes public-key-bytes)
-     true
-     (catch RuntimeException _ false)))
-  ([signature body timestamp public-key charset-name]
-   (if-let-all [signature-bytes (cond-> signature (not (bytes? signature)) hex->bytes)
-                public-key-bytes (cond-> public-key (not (bytes? public-key)) hex->bytes)
-                ^bytes body-bytes (cond-> body (string? body) (encode body charset-name))
-                ^bytes timestamp-bytes (cond-> timestamp (string? timestamp) (encode timestamp charset-name))
-                message-bytes (byte-array (+ (alength timestamp-bytes) (alength body-bytes)))]
-     (do
-       (System/arraycopy timestamp-bytes 0 message-bytes 0 (alength timestamp-bytes))
-       (System/arraycopy body-bytes 0 message-bytes (alength timestamp-bytes) (alength body-bytes))
-       (authentic? signature-bytes message-bytes public-key-bytes))
-     false)))
```
The contract and signature of `authentic?` should be kept. It allows for more parametrisation and accepts a greater variety of types.

rafaeldelboni replied:
My problem with the current signature is the `System/arraycopy` you need to make to combine the two byte arrays, if you had them as a string before they are bytes you can just merge with an str function and then convert to a byte array which makes the code easier to read.

JohnnyJayJay replied:
Since the verification uses byte arrays in the end, that is what `authentic?` coerces to. This has 3 benefits:
- It's fast
- It saves unnecessary (and possibly error-prone) conversions
- It allows for different combinations of input types without additional overhead

It may not be the nicest to read, but performance was one thing I wanted to focus on right away.

rafaeldelboni replied:
:+1:

rafaeldelboni commented:
> add new functions like new-verifier and helper functions for tests (sign, generate-keypair, ...) where needed

Do you think those should be in a separate test helper file?

rafaeldelboni commented:
@JohnnyJayJay can you check again, solved all your points, and added more unit tests to cover all `authentic?` scenarios.

JohnnyJayJay commented on src/ring_discord_auth/validation.clj:
```diff
@@ -0,0 +1,181 @@
+                [(hex (bit-shift-right v 4)) (hex (bit-and v 0x0F))]))]
+      (str/join (mapcat hexify-byte byte-array)))))
+
+(defn generate-keypair
+  "Generate Ed25519 key pair.
+
+  Return {:private `Ed25519PrivateKeyParameters`
+          :public `Ed25519PublicKeyParameters`}"
+  []
+  (let [random (SecureRandom.)
+        kpg    (Ed25519KeyPairGenerator.)
+        _ (.init kpg (Ed25519KeyGenerationParameters. random))
+        key-pair (.generateKeyPair kpg)]
+    {:private (cast Ed25519PrivateKeyParameters (.getPrivate key-pair))
+     :public  (cast Ed25519PublicKeyParameters (.getPublic key-pair))}))
+
+(defn new-signer
+  "Return new instance of `Ed25519Signer` initialized by private key"
+  [private-key]
+  (let [signer (Ed25519Signer.)]
+    (.init signer true private-key)
+    signer))
+
+(defn sign
+  "Generate signature for msg byte array.
+
+  Return byte array with signature."
+  [^Ed25519Signer signer msg-bytes]
+  (.update signer msg-bytes 0 (alength msg-bytes))
+  (.generateSignature signer))
```
I feel like these functions should be part of the test namespace since they are not needed by users of this library. Feel free to correct me on that

rafaeldelboni replied:
:+1:

JohnnyJayJay commented on src/ring_discord_auth/validation.clj:
```diff
@@ -0,0 +1,181 @@
+    (.init signer false public-key)
+    signer))
+
+(defn verify
+  "Verify signature for msg byte array.
+
+  Returns `true` if valid signature and `false` if not."
+  [^Ed25519Signer signer msg-bytes signature]
+  (.update signer msg-bytes 0 (alength msg-bytes))
+  (.verifySignature signer signature))
+
+(defn verify-request
+  "Verify discord payload with app public-key, request body, signature and timestamp headers
+
+  Returns `true` if the message is authentic, `false` if not."
+  [public-key timestamp body signature]
+  (verify (new-verifier (Ed25519PublicKeyParameters. (hex->bytes public-key) 0))
+          (.getBytes (str timestamp body) "utf8")
+          (hex->bytes signature)))
```
Could you remove these two functions and instead add support to pass an `Ed25519Signer` to `authentic?`? 
I.e. the signature of authentic? would become `[signature-bytes message-bytes public-key-bytes-or-verifier]` and same for the other arity. That would remove potential confusion as to why there are two functions that essentially do the same thing

rafaeldelboni replied:
removed `verify-request` :+1:

JohnnyJayJay commented on src/ring_discord_auth/validation.clj:
```diff
@@ -0,0 +1,181 @@
+          (.getBytes (str timestamp body) "utf8")
+          (hex->bytes signature)))
+
+(defn authentic?
+  "Checks whether a signature is authentic, given a message and a public key.
+
+  This function has 2 arities: a general purpose one that simply delegates to a crypto library and one tailored to what Discord provides.
+
+  The general purpose arity takes a signature, a message and a public key (all of which must be byte arrays) and checks the authenticity.
+  The other arity takes a signature (byte array or hex string), public key (byte array or hex string), body (byte array or string), timestamp (byte array or string).
+
+  It combines timestamp and body and uses that as the message.
+
+  Returns `true` if the message is authentic, `false` if not."
+  ([signature-bytes message-bytes public-key-bytes]
+   (verify (new-verifier (Ed25519PublicKeyParameters. public-key-bytes 0))
+            message-bytes
+            signature-bytes))
+  ([signature body timestamp public-key charset-name]
+   (if-let-all [signature-bytes (cond-> signature (not (bytes? signature)) hex->bytes)
+                public-key-bytes (cond-> public-key (not (bytes? public-key)) hex->bytes)
+                ^bytes body-bytes (cond-> body (string? body) (encode charset-name))
+                ^bytes timestamp-bytes (cond-> timestamp (string? timestamp) (encode charset-name))
+                message-bytes (byte-array (+ (alength timestamp-bytes) (alength body-bytes)))]
+     (do
+       (System/arraycopy timestamp-bytes 0 message-bytes 0 (alength timestamp-bytes))
+       (System/arraycopy body-bytes 0 message-bytes (alength timestamp-bytes) (alength body-bytes))
+       (authentic? signature-bytes message-bytes public-key-bytes))
+     false)))
```
Regarding my comment above: a suggestion for `authentic?`
```suggestion
(defn authentic?
  "Checks whether a signature is authentic, given a message and a public key.

  This function has 2 arities: a general purpose one that simply delegates to a crypto library and one tailored to what Discord provides.

  The general purpose arity takes a signature, a message and a public key (all of which must be byte arrays - the public key may also be a verifier obtained by [[new-verifier]]) and checks the authenticity.
  The other arity takes a signature (byte array or hex string), public key (byte array, hex string or verifier), body (byte array or string), timestamp (byte array or string).

  It combines timestamp and body and uses that as the message.

  Returns `true` if the message is authentic, `false` if not."
  ([signature-bytes message-bytes public-key-bytes-or-verifier]
   (let [verifier (cond-> public-key-bytes-or-verifier  (bytes? public-key-bytes-or-verifier)  new-verifier)]
     (.update verifier message-bytes 0 (alength message-bytes))
     (.verifySignature verifier signature-bytes)))
  ([signature body timestamp public-key charset-name]
   (if-let-all [signature-bytes (cond-> signature (not (bytes? signature)) hex->bytes)
                public-key-or-verifier (cond-> public-key (string? public-key) hex->bytes)
                _ (or (bytes? public-key-or-verifier) (instance? Ed25519Signer public-key-or-verifier))
                ^bytes body-bytes (cond-> body (string? body) (encode charset-name))
                ^bytes timestamp-bytes (cond-> timestamp (string? timestamp) (encode charset-name))
                message-bytes (byte-array (+ (alength timestamp-bytes) (alength body-bytes)))]
     (do
       (System/arraycopy timestamp-bytes 0 message-bytes 0 (alength timestamp-bytes))
       (System/arraycopy body-bytes 0 message-bytes (alength timestamp-bytes) (alength body-bytes))
       (authentic? signature-bytes message-bytes public-key-or-verifier))
     false)))
```

JohnnyJayJay reviewed (changes requested):
Looks good :+1: 
Left two remarks concerning namespace pollution and one possible solution for one of them.

Also, still an open question whether validation logic should be moved to `core` and the middleware should be moved to `ring`. I feel like that'd be a better way of splitting it.

rafaeldelboni commented:
Hey, I changed the namespaces as you suggested and added an arity version to `authentic?` to also receive a `Ed25519Signer`, but I made a breaking change for the sake of code readability, the `authentic?` function was getting to big so I split it up a little bit.

And added even more unit tests.

JohnnyJayJay reviewed (approved):
:+1: 

Will check this again locally tomorrow and then merge if there are no further complications.

JohnnyJayJay commented:
Looks good, only a few reflection warnings left. I'll take care of them in develop.
